### DISCUSSION TITLE
x264/x265 licensing

### DISCUSSION THREAD

**Discussion by alexthedamager** [OvenMediaLabs/OvenMediaEngine#1333] [Ideas] [answered]
In #869 there was talk of not being able to use x264 / x265 in OME due to licensing conflicts, wherein it was suggested that x264 and x265 are GPLv2 only rather than "GPLv2 or later". 

I wanted to share some information in case it's useful. It would appear that both x264 and x265 are actually "GPLv2 or later", rather than "GPLv2 only". That being the case, I believe it is okay from a licensing perspective to use both x264 and x265 in OME. OME is AGPLv3. 

### **x264** 
- x264 source code [here](https://code.videolan.org/videolan/x264/-/blob/master/COPYING) which states in the COPYING file and in the .h and .cpp files:
> "This program is free software; you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation; **either version 2 of the License, or (at your option) any later version.**"
- x264 issue in which it is suggested to ammend the x264 website to reflect "GPLv2 or later" [here](https://code.videolan.org/videolan/x264/-/issues/40)
- Wikipedia page stating "GPL-2.0-or-later" [here](https://en.wikipedia.org/wiki/X264)

### **x265**
- FFMpeg discussion thread in which it is established that x265 is "GPLv2 or later" [here](https://trac.ffmpeg.org/ticket/9466)
- x265 source code [here](https://bitbucket.org/multicoreware/x265_git/src/master/COPYING) which states in the COPYING file and in the .h and .cpp files:
> "This program is free software; you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation; **either version 2 of the License, or (at your option) any later version.**"
- x265 issue in which it is suggested to ammend the x265 website to reflect "GPLv2 or later" [here](https://bitbucket.org/multicoreware/x265_git/issues/607/license-faq-enchancement-suggestion)
- Wikipedia page stating "GPL-2.0-or-later" [here](https://en.wikipedia.org/wiki/X265)

**GNU Compatability Matrix:**
Finally, the GNU Compatability matrix can be found [here](https://www.gnu.org/licenses/gpl-faq.en.html) and would indicate that "GPLv2 or later" is compatible with GPLv3 
![Screenshot 2023-08-05 at 12 51 06](https://github.com/AirenSoft/OvenMediaEngine/assets/67900507/05b35b57-4ea6-453a-a293-7d6243cf4b68)

I might be missing something, but figured this might be important information for the community to be aware of, and might enable OME to make use of x264 and x265 encoders for live transcoding.

**Comment by dimiden** [comment]
Thank you for the detailed clarification.
However, as far as I know, GPLv2 or later is unfortunately not compatible with AGPLv3. Therefore, it seems that we still cannot use x264/x265.


Disclaimer - I am not a licensing expert, so I may not be accurate. If I'm mistaken about this, please let me know. :)

**Comment by getroot** [comment]
I am also not someone who can give legal advice on open source licenses. But I read the FAQ URL below and thought that as long as OME maintains AGPLv3, GPLv2+ cannot be combined.
https://www.gnu.org/licenses/license-list.html#AGPLv3.0

**Comment by alexthedamager** [comment]
The FAQ section mentions "Please note that the GNU AGPL is not compatible with GPLv2". This is true, it is not compatible with GPLv2, but I believe the crux of the matter is that you can treat x264 and x265 as GPLv3 (because their licenses are "GPLv2 or later") and in doing so, can use them in an AGPLv3 project like OME:

- x264/x265 is "GPLv2 or later": This means you can choose to use x264/x265 under either GPLv2 or any later version of the GPL, including GPLv3.
- AGPLv3 allows combination with GPLv3: The AGPLv3 license has provisions (in Section 13) that allow you to combine it with GPLv3 code.
- Combine them under GPLv3: Since x264/x265 can be treated as GPLv3 due to its "or later" clause, and AGPLv3 allows combination with GPLv3, you can use x264/x265 in an AGPLv3 project by considering x264/x265 as GPLv3.

The part of the FAQ that @getroot referred to states:
> Please note that the GNU AGPL is not compatible with GPLv2. It is also technically not compatible with GPLv3 in a strict sense: you cannot take code released under the GNU AGPL and convey or modify it however you like under the terms of GPLv3, or vice versa. 

I believe this means that whilst you can't take AGPLv3 code and directly relicense or distribute it as GPLv3 (or vice versa), you can have AGPLv3 and GPLv3 side-by-side in a project. 

Additionally, on the GNU website above the compatability matrix [screenshot in my previous comment](https://github.com/AirenSoft/OvenMediaEngine/discussions/1333#discussion-5481095), it states "Each place that the matrix states GPLv3, the same statement about compatibility is true for AGPLv3 as well." 

I am not a lawyer, but I do think this is a way to include x264/x265 in OME whilst maintaining AGPLv3.

**Reply by jbkempf** [reply]
My reading is similar to @llspalex.

Use x264 as GPLv3 and use the article 13 to use GPLv3 code inside AGPLv3. You cannot give it back, though, and I would argue that you should modularize as much as possible the x264 module.

For x265, I don't know yet.

**Comment by basisbit** [comment]
Might be helpful in the considerations here: I am still using a 1+ year old OME codebase where ffmpeg is used for encoding instead of openh264, because of the result of ffmpeg encoder being a *LOT* better quality at similar settings when streaming content with music visualizations that have lots of bright -> dark -> bright changes and lots of fragments while using quite low video bitrates (1.1Mb/s) and quite low resolution (360p). ffmpeg encoder gives a good result even at preset faster, while the result of openh264 is what I would call potato-quality even at preset medium, while using same streamed video and same other settings.

Would be cool to be able to maybe specify different licensed-build-targets which then use the best allowed encoder/decoder for it, so that someone who is not a big company can still use ffmpeg, while a big company would for example have to pay license fees or use the free/open encoder which has lower encoding quality.

**Comment by Baa14453** [comment, marked as answer]
For some added clarification, I contacted VideoLan for their thoughts and Jean provided the below

> Hello,
> 
> x264 is either licensable under GPLv2 or GPLv3.
> GPLv3 can be used in AGPLv3 projects, but not the other way around.
> 
> So, I think this is fine and that:
> https://github.com/AirenSoft/OvenMediaEngine/discussions/1333#discussioncomment-6667837
> is correct for x264.
> 
> I'm not sure this applied to x265 though.
> 
> -- 
> Jean-Baptiste Kempf -  President

So, from VideoLan's perspective, OME are fine to use libx264

**Reply by getroot** [reply]
Thanks for clarifying. I'll re-enable x264 soon

**Comment by hashworks** [comment]
What is required to re-introduce x264? I guess reversing 2044f7e76d079fc60fcacbcefa250fe97b8ad744 and updating the versions wouldn't be enough?

**Reply by getroot** [reply]
The only thing that is needed is @Keukhan 's time. He will do this task soon.
